src/Microsoft.ApiDesignGuidelines.Analyzers/Core/IdentifiersShouldHaveCorrectSuffix.cs

```diff
@@ -25,24 +29,141 @@ public abstract class IdentifiersShouldHaveCorrectSuffixAnalyzer : DiagnosticAna
+            var interfaceTypeSuffixMap = new Dictionary<INamedTypeSymbol, SuffixInfo>();
+
+            foreach (var tuple in s_baseTypesAndTheirSuffix)
+            {
+                var wellKnownNamedType = context.Compilation.GetTypeByMetadataName(tuple.Item1);
+
+                if (wellKnownNamedType != null && wellKnownNamedType.OriginalDefinition != null)
+                {
+                    // If the type is interface
+                    if (wellKnownNamedType.OriginalDefinition.TypeKind == TypeKind.Interface)
+                    {
+                        interfaceTypeSuffixMap.Add(wellKnownNamedType.OriginalDefinition, SuffixInfo.Create(tuple.Item2, tuple.Item3));
+                    }
+                    else
+                    {
+                        baseTypeSuffixMap.Add(wellKnownNamedType.OriginalDefinition, SuffixInfo.Create(tuple.Item2, tuple.Item3));
+                    }
+                }
+            }
+
+            if (baseTypeSuffixMap.Count > 0 || interfaceTypeSuffixMap.Count > 0)
+            {
+                context.RegisterSymbolAction((saContext) =>
+                {
+                    var namedTypeSymbol = saContext.Symbol as INamedTypeSymbol;
+                    if (namedTypeSymbol.GetResultantVisibility() != SymbolVisibility.Public)
+                    {
+                        return;
+                    }
+
+                    var baseType = namedTypeSymbol.GetBaseTypes().FirstOrDefault(bt => baseTypeSuffixMap.Keys.Contains(bt.OriginalDefinition));
+                    if (baseType != null)
+                    {
+                        var suffixInfo = baseTypeSuffixMap[baseType.OriginalDefinition];
+                        var rule = suffixInfo.CanSuffixBeCollection ? SpecialCollectionRule : DefaultRule;
+                        if ((suffixInfo.CanSuffixBeCollection && !namedTypeSymbol.Name.EndsWith(CollectionString) && !namedTypeSymbol.Name.EndsWith(suffixInfo.Suffix)) ||
+                            (!suffixInfo.CanSuffixBeCollection && !namedTypeSymbol.Name.EndsWith(suffixInfo.Suffix)))
+                        {
+                            saContext.ReportDiagnostic(namedTypeSymbol.CreateDiagnostic(rule, namedTypeSymbol.ToDisplayString(), suffixInfo.Suffix));
+                        }
+
+                        return;
+                    }
+
+                    var implementedInterface = namedTypeSymbol.AllInterfaces.FirstOrDefault(i => interfaceTypeSuffixMap.Keys.Contains(i.OriginalDefinition));
+                    if (implementedInterface != null)
+                    {
+                        var suffixInfo = interfaceTypeSuffixMap[implementedInterface.OriginalDefinition];
+                        if (!namedTypeSymbol.Name.EndsWith(suffixInfo.Suffix))
+                        {
+                            saContext.ReportDiagnostic(namedTypeSymbol.CreateDiagnostic(DefaultRule, namedTypeSymbol.ToDisplayString(), suffixInfo.Suffix));
```

Comment:
@mavasani I looked at the old implementation and there is an ordering in what they intended to check first. They initially checked against the BaseTypes which we do as well. We check against the Base Types before moving to the interface Types. The old implementation however has some interfaces also mentioned in their BaseTypesList :) and then they check for collection based types. 

In general, there is no specified ordering in the old implementation. So the current implementation looks for matches against the basetypes and if there isn't any then we move to interface types.  This sounds like a reasonable approach.

Also the case of having mentioned 2 interfaces is a corner case. For Eg: How often are we going to see a type that has IPermission and ICollection implemented? If the users have issue with it then we can always revisit to decide the correct behavior.